server/vulnerabilities/cpe.go

```diff
@@ -139,67 +144,312 @@ func cleanAppName(appName string) string {
 	return strings.TrimSuffix(appName, ".app")
 }
 
-var onlyAlphaNumeric = regexp.MustCompile("[^a-zA-Z0-9]+")
-
-func CPEFromSoftware(db *sqlx.DB, software *fleet.Software) (string, error) {
-	targetSW := ""
-	switch software.Source {
-	case "apps":
-		targetSW = "macos"
-	case "python_packages":
-		targetSW = "python"
-	case "chrome_extensions":
-		targetSW = "chrome"
-	case "firefox_addons":
-		targetSW = "firefox"
-	case "safari_extensions":
-		targetSW = "safari"
-	case "deb_packages":
-	case "portage_packages":
-	case "rpm_packages":
-	case "npm_packages":
-		targetSW = `"node.js"`
-	case "atom_packages":
-	case "programs":
-		targetSW = `"windows*"`
-	case "ie_extensions":
-	case "chocolatey_packages":
+var onlyAlphaNumeric = regexp.MustCompile(`[^a-zA-Z0-9]+`)
+
+// sanitizeMatch sanitizes the search string for sqlite fts queries. Replaces all special characters with spaces.
+func santizeMatch(s string) string {
+	return onlyAlphaNumeric.ReplaceAllString(s, " ")
+}
+
+var sanitizeVersionRe = regexp.MustCompile(`[^a-zA-Z0-9_-]+`)
+
+// sanitizeVersion attempts to sanitize versions and attempt to make it dot separated.
+// Eg Zoom reports version as "5.11.1 (8356)". In the NVD CPE dictionary it should be 5.11.1.8356.
+func sanitizeVersion(version string) string {
+	parts := onlyAlphaNumeric.Split(version, -1)
+	return strings.Trim(strings.Join(parts, "."), ".")
+}
+
+// TODO: add more vendors
+var macOSVendors = map[string]string{
+	"com.postmanlabs.mac":           "getpostman",
+	"org.virtualbox.app.VirtualBox": "oracle",
+}
+
+const cpeTranslationsFilename = "cpe_translations.json"
+
+func loadCPETranslations(path string) (CPETranslations, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	var translations CPETranslations
+	if err := json.NewDecoder(f).Decode(&translations); err != nil {
+		return nil, fmt.Errorf("decode json: %w", err)
+	}
+
+	return translations, nil
+}
+
+func DownloadCPETranslations(vulnPath string, client *http.Client) error {
+	ghClient := github.NewClient(client)
+
+	opts := &github.RepositoryContentGetOptions{
+		Ref: "michal-6628-macos-vuln",
+	}
+	r, resp, err := ghClient.Repositories.DownloadContents(context.Background(), "fleetdm", "nvd", cpeTranslationsFilename, opts)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("github response non 200 status code: %d", resp.StatusCode)
+	}
+
+	tmpFile, err := os.CreateTemp(vulnPath, cpeTranslationsFilename)
+	if err != nil {
+		return fmt.Errorf("create temporary file: %w", err)
+	}
+	defer tmpFile.Close()
+
+	// Clean up tmp file if not moved
```

Comment:
curious: is there a reason for creating a temp file and copying that instead of directly using `os.Create` and writing the contents in the final destination? I can see the `moved` variable being confusing as the time passes and this functionality grows

Reply:
Some of this code was copied from `pkg/download/download.go`. The reason for doing it this way is in case the write fails. If the file exists, `os.Create` will truncate the file. Writing to a temporary file and then renaming it ensures it is an atomic operation.